python/rmm/rmm/statistics.py

```diff
@@ -0,0 +1,350 @@
+        ----------
+        num_calls
+            Number of times this code block was invoked.
+        memory_total
+            Total number of bytes allocated.
+        memory_peak
+            Peak number of bytes allocated.
+        """
+
+        num_calls: int = 0
+        memory_total: int = 0
+        memory_peak: int = 0
+
+        def add(self, memory_total: int, memory_peak: int):
+            self.num_calls += 1
+            self.memory_total += memory_total
+            self.memory_peak = max(self.memory_peak, memory_peak)
+
+    def __init__(self) -> None:
+        self._lock = threading.Lock()
+        self._records: Dict[str, ProfilerRecords.MemoryRecord] = defaultdict(
+            ProfilerRecords.MemoryRecord
+        )
+
+    def add(self, name: str, data: Statistics) -> None:
+        """Add memory statistics to the record named `name`.
+
+        This method is thread-safe.
+
+        Parameters
+        ----------
+        name
+            Name of the record.
+        data
+            Memory statistics of `name`.
+        """
+        with self._lock:
+            self._records[name].add(
+                memory_total=data.total_bytes, memory_peak=data.peak_bytes
+            )
+
+    @property
+    def records(self) -> Dict[str, MemoryRecord]:
+        """Dictionary mapping record names to their memory statistics."""
+        return dict(self._records)
+
+    def report(
+        self,
+        ordered_by: Literal[
+            "num_calls", "memory_peak", "memory_total"
+        ] = "memory_peak",
+    ) -> str:
+        """Pretty format the recorded memory statistics.
+
+        Parameters
+        ----------
+        ordered_by
+            Sort the statistics by this attribute.
+
+        Return
+        ------
+        The pretty formatted string of the memory statistics
+        """
+
+        # Sort by `ordered_by`
+        records = sorted(
+            ((name, data) for name, data in self.records.items()),
+            key=lambda x: getattr(x[1], ordered_by),
+            reverse=True,
+        )
+        ret = "Memory Profiling\n"
+        ret += "================\n\n"
+        if len(records) == 0:
+            return ret + "No data, maybe profiling wasn't enabled?"
+        ret += (
+            "Legends:\n"
+            "  ncalls       - number of times the function or code block "
+            "was called\n"
+            "  memory_peak  - peak memory allocated in function or code "
+            "block (in bytes)\n"
+            "  memory_total - total memory allocated in function or code "
+            "block (in bytes)\n"
+        )
+        ret += f"\nOrdered by: {ordered_by}\n"
+        ret += "\nncalls     memory_peak    memory_total  "
+        ret += "filename:lineno(function)\n"
+        for name, data in records:
+            ret += f"{data.num_calls:6,d} {data.memory_peak:15,d} "
+            ret += f"{data.memory_total:15,d}  {name}\n"
+        return ret[:-1]  # Remove the final newline
```

Comment:
*suggestion*:

```
ret += "\n".join(f"{data.num_calls:6,d} {data.memory_peak:15,d} {data.memory_total:15,d} {name}" for name, data in records)
```

Reply:
The line has to be broken up because of the line width limit, which makes the code less clear IMO.